Move Key interning to rust

stuhood opened the pull request:
### Problem

Key interning was implemented on the python side of the FFI boundary, which meant spending a longer time on the python side than was strictly necessary in the `key_for` and `val_for` externs.

### Solution

Move the Key intern map and the `key_for`/`val_for` functions to rust such that they can spend less time with the GIL acquired, and avoid a few clones.

### Result

`./pants list ::` is about ~8% faster in a large repository. No change in memory usage.

illicitonion commented:
Please add _all_ the documentation to this corner of the world. In particular, a paragraph of comments in either the python or the rust (and a pointer from the other) which explains an entire "I have an object, serialize it, pass it across the boundary, and use it" flow, with reference to the functions called, and the intermediate types, would be really useful.

Specific things which confused me when I was reading the code:
 * What's an `InternKey` (it's undocumented)? What's the difference between a `Key` and an `InternKey`? How do their namespaces differ?
 * What's the lifetime of things which have been interned? Is this how we intend it to be forever, or a stop-gap until we work out how to change that?

stuhood commented:
Rebased, and added two commits that 1) move `struct Interns` and its private helpers into their own module to make it more obvious that only `Interns` is public, 2) add documentation of interning and the relationship between `Key` and `Value`.

illicitonion commented on src/rust/engine/src/interning.rs:
```diff
@@ -0,0 +1,86 @@
+  pub fn insert(&mut self, v: Value) -> Key {
+    let ident = externs::identify(&v);
+    let type_id = ident.type_id;
+    let mut maybe_id = self.id_generator;
```
Why do you need `maybe_id`? Why can't you just use `self.id_generator` everywhere? I would rather have a bool tracking whether the insert happened, than need to reason about changes to copies of struct state...

stuhood replied:
Because `self(.id_generator)` can't be borrowed immutably while it is borrowed mutably for the write. I think I can move a copy into the closure though.

illicitonion commented on src/rust/engine/src/interning.rs:
```diff
@@ -0,0 +1,86 @@
+#[derive(Default)]
+pub struct Interns {
+  forward: HashMap<InternKey, Key, FNV>,
+  reverse: HashMap<Id, Value, FNV>,
```
How expensive is it to have these actually be inverses of each other (`HashMap<InternKey, Key>` and `Hashmap<Key, InternKey>)`? I think it would be easier to reason about the interning if things didn't implicitly unwrap to `Id` (which has many semantic purposes), and instead kept their associated type metadata...

stuhood replied:
Yep, looks fine.

illicitonion commented on src/rust/engine/src/interning.rs:
```diff
@@ -0,0 +1,86 @@
+///
+#[derive(Default)]
+pub struct Interns {
+  forward: HashMap<InternKey, Key, FNV>,
```
Why does this use `Key` as the value type, rather than `Id`, given only the `Id` is ever actually used?

stuhood replied:
The `TypeId` of the `Key` is used as well (in `nodes.rs`). Unclear whether `TypeId` really needs to be part of the key (we could call an extern explicitly when we need it), but we use it at the beginning of every `Select` call, and we make multiple `Selects` per key.

stuhood commented on src/python/pants/engine/native.py:
```diff
@@ -758,4 +725,5 @@ def tc(constraint):
 
   def set_panic_handler(self):
-    self.lib.set_panic_handler()
+    # self.lib.set_panic_handler()
```
Yep, fixed.

stuhood commented on src/rust/engine/hashing/src/lib.rs:
```diff
@@ -141,40 +141,42 @@ mod fingerprint_tests {
-        0xab,
-        0xab,
-      ]),
+      Fingerprint::from_bytes_unsafe(
```
rustfmt reformatted this when I switched from `&vec![..]` to `&[..]`.

stuhood commented on src/rust/engine/src/externs.rs:
```diff
@@ -182,9 +178,11 @@ pub fn call(func: &Value, args: &[Value]) -> Result<Value, Failure> {
-  call(&val_for_id(func.0), args).unwrap_or_else(|e| {
-    panic!("Core function `{}` failed: {:?}", id_to_str(func.0), e);
+pub fn unsafe_call(func: &Function, args: &[Value]) -> Value {
+  let interns = INTERNS.read().unwrap();
```
In order to avoid cloning the value, yea: it needs to be held. It's a tradeoff: `Value::clone()` isn't cheap, and so we're avoiding an extern call by holding a (shared) lock longer.

stuhood commented on src/rust/engine/src/rule_graph.rs:
```diff
@@ -675,8 +671,11 @@ pub fn type_str(type_id: TypeId) -> String {
+    externs::id_to_str(type_id.0)
+    // TODO: We should be calling `__name__` here, but we're not. Would need to get
+    // a Value for the TypeId.
+    // let as_val =
```
Thanks. This was also the reason for (most of?) the failing tests.
